KEP-2170: Add the manifests overlay for Kubeflow Training V2

Doris-xm opened the pull request:
**What this PR does / why we need it**:

This PR adds the manifests overlay for Kubeflow Training V2, allowing to install it within Kubeflow Platform.

**Which issue(s) this PR fixes** :
Fixes #2381 

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/training/) included if any changes are user facing

Commit:
Add the manifests overlay for Kubeflow Training V2

Signed-off-by: Xinmin Du <10803082+doris-xm@user.noreply.gitee.com>
Signed-off-by: Xinmin Du <2812493086@qq.com>

Electronic-Waste commented:
Will review it later.

/cc @kubeflow/wg-training-leads @kubeflow/release-team @saileshd1402 
/ok-to-test
/rerun-all

Electronic-Waste commented on manifests/v2/overlays/kubeflow/kustomization.yaml:
```diff
@@ -0,0 +1,15 @@
+resources:
+  - ../../base
+  - kubeflow-training-roles.yaml
+  - https://github.com/kubernetes-sigs/jobset/releases/download/v0.6.0/manifests.yaml
```
Please add comments to make this item consistent with: https://github.com/kubeflow/training-operator/blob/be2e29effb63539c9abc94b26adc43b2f256a842/manifests/v2/overlays/standalone/kustomization.yaml#L10-L11

andreyvelich replied:
I am wondering for Kubeflow Platform installation, should we actually deploy JobSet controller in the `kubeflow` namespace ?
@kannon92 also made it possible: https://github.com/kubernetes-sigs/jobset/pull/719
cc @kubeflow/wg-manifests-leads

Electronic-Waste commented on manifests/v2/overlays/kubeflow/kubeflow-training-roles.yaml:
```diff
@@ -0,0 +1,91 @@
+metadata:
+  name: kubeflow-training-admin-v2
+  labels:
+    rbac.authorization.kubeflow.org/aggregate-to-kubeflow-admin: "true"
```
This label will conflict with:

https://github.com/kubeflow/training-operator/blob/be2e29effb63539c9abc94b26adc43b2f256a842/manifests/overlays/kubeflow/kubeflow-training-roles.yaml#L6

Please rename it to `rbac.authorization.kubeflow.org/aggregate-to-kubeflow-admin-v2` to avoid conflicts.

Electronic-Waste commented on manifests/v2/overlays/kubeflow/kubeflow-training-roles.yaml:
```diff
@@ -0,0 +1,91 @@
+aggregationRule:
+  clusterRoleSelectors:
+    - matchLabels:
+        rbac.authorization.kubeflow.org/aggregate-to-kubeflow-training-admin: "true"
```
Same as above.

Electronic-Waste commented on manifests/v2/overlays/kubeflow/kubeflow-training-roles.yaml:
```diff
@@ -0,0 +1,91 @@
+metadata:
+  name: kubeflow-training-edit-v2
+  labels:
+    rbac.authorization.kubeflow.org/aggregate-to-kubeflow-edit: "true"
+    rbac.authorization.kubeflow.org/aggregate-to-kubeflow-training-admin: "true"
```
Same as above.

Electronic-Waste commented on manifests/v2/overlays/kubeflow/kubeflow-training-roles.yaml:
```diff
@@ -0,0 +1,91 @@
+metadata:
+  name: kubeflow-training-view-v2
+  labels:
+    rbac.authorization.kubeflow.org/aggregate-to-kubeflow-view: "true"
```
Same as above.

Electronic-Waste commented on manifests/v2/overlays/kubeflow/kubeflow-training-roles.yaml:
```diff
@@ -0,0 +1,91 @@
+      - get
+      - list
+      - watch
+  - apiGroups:
+      - ""
+    resources:
+      - events
+    verbs:
+      - get
+      - list
+      - watch
```
I haven't seen any event resources in Training V2. Maybe we should remove them

https://github.com/kubeflow/training-operator/blob/be2e29effb63539c9abc94b26adc43b2f256a842/pkg/controller.v2/trainjob_controller.go#L69-L72

https://github.com/kubeflow/training-operator/blob/be2e29effb63539c9abc94b26adc43b2f256a842/pkg/controller.v1/pytorch/pytorchjob_controller.go#L106-L114

Electronic-Waste replied:
PTAL👀 @kubeflow/wg-training-leads. Do you think we should reserve `events` resource in the manifests?

Doris-xm replied:
Thanks for your review! It seems `event` resources are required for `EventRecorder` used in `TrainJobReconciler` to log events related to training jobs.

> https://github.com/kubeflow/training-operator/blob/be2e29effb63539c9abc94b26adc43b2f256a842/pkg/controller.v2/trainjob_controller.go#L53-L58

However, the `event` resource permissions are not included in the `kubebuilder` annotations in the `trainjob_controller.go` 
> https://github.com/kubeflow/training-operator/blob/be2e29effb63539c9abc94b26adc43b2f256a842/pkg/controller.v2/trainjob_controller.go#L69-L71

I'm not sure I understand it correctly. Appreciate your clarification.

Electronic-Waste replied:
SGTM! However, in controller v2, we haven't called `r.recorder.Eventf()` yet. Maybe we could add `events` resource's permission to the Kubeflow Manifests when we call `r.recorder.Eventf()`.

andreyvelich replied:
Actually, events will be required in the SDK, like for V1: https://github.com/kubeflow/training-operator/blob/master/sdk/python/kubeflow/training/api/training_client.py#L1309
We haven't implemented it yet in V2, but we will do it in the future.

Electronic-Waste commented on manifests/v2/overlays/kubeflow/kubeflow-training-roles.yaml:
```diff
@@ -0,0 +1,91 @@
+  - apiGroups:
+      - kubeflow.org
+    resources:
+      - clustertrainingruntimes
+      - trainingruntimes
```
We could also edit `trainingruntimes` and `clustertrainingruntimes` in the Kubeflow platform. I think, their perssmission granularity should be the same with `trainjobs`:)

Electronic-Waste reviewed (commented):
Thanks for creating this @Doris-xm! I left some initial comments for you.

Btw, I recommend that you could learn more about the concept of Training V2. This will help you update the manifests overlay in training-operator and kubeflow/manifests:)

FYI: [KubeCon 2024 NA Talk](https://kccncna2024.sched.com/event/1i7nV?iframe=no) by Andrey and Yuki

Electronic-Waste commented:
/rerun-all

Doris-xm commented on manifests/v2/overlays/kubeflow/kubeflow-training-roles.yaml:
```diff
@@ -0,0 +1,91 @@
+    resources:
+      - trainjobs/status
+    verbs:
+      - get
```
Thanks for your careful review！

Commit:
Update manifest: adjust permissions, and format changes

Signed-off-by: Xinmin Du <10803082+doris-xm@user.noreply.gitee.com>
Signed-off-by: Xinmin Du <2812493086@qq.com>

Electronic-Waste reviewed (approved):
Basically LGTM. Thank you for your great contributions!

/lgtm
/assign @kubeflow/wg-training-leads @kubeflow/wg-manifests-leads

andreyvelich commented on manifests/v2/overlays/kubeflow/kustomization.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Can we name this overlay: `kubeflow-platform` to make it consistent with the naming: https://www.kubeflow.org/docs/started/introduction/#what-is-kubeflow-platform.

andreyvelich reviewed (commented):
Thank you for this great contribution @willb!
/assign @kubeflow/wg-manifests-leads @kubeflow/release-team

Commit:
Update manifest: rename overlay, adjust event permissions

Signed-off-by: Xinmin Du <10803082+doris-xm@user.noreply.gitee.com>
Signed-off-by: Xinmin Du <2812493086@qq.com>

Electronic-Waste commented on manifests/v2/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,20 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: kubeflow
```
May I ask a question? Since we have defined the namespace for manager and webhook:

https://github.com/kubeflow/training-operator/blob/1dfa40c12516fc9eb2ce12c5ef52da7d46670457/manifests/v2/base/manager/kustomization.yaml#L3-L4

Does `kubeflow` namespace take any effect here? WDYT👀 @kubeflow/wg-training-leads  @Doris-xm

Doris-xm replied:
Thanks for your reviewing.  It seems that `namespace: kubeflow` in the overlay won't override `namespace: kubeflow-system` defined for manager and webhook.   I'm wondering if we should unify the namespace as `kubeflow` when installing them within Kubeflow Platform, aligning with other modules.

andreyvelich replied:
I think, we should make our namespace configurable and use `kubeflow` namespace for Kubeflow Platform installation for now.
By default the Kubeflow Training V2 will use the `kubeflow-system` namespace.

Electronic-Waste replied:
@Doris-xm  FYI, I think what @andreyvelich means is that we should address these `TODO`:

https://github.com/kubeflow/training-operator/blob/1dfa40c12516fc9eb2ce12c5ef52da7d46670457/manifests/v2/base/manager/kustomization.yaml#L3-L4

https://github.com/kubeflow/training-operator/blob/1dfa40c12516fc9eb2ce12c5ef52da7d46670457/manifests/v2/base/webhook/kustomization.yaml#L13-L14

And specify their namespace in `overlays` and `kubeflow-platform` dir. Because we need to make namesapce configurable and install these components in different namspace:

1. Standalone `training-operator`: Install all components in `kubeflow-system` namespace
2. Kubeflow Platform: Install all components in `kubeflow` namespace

Doris-xm replied:
Thanks for your detailed explanation! I've moved the namespace configuration from `base` to `overlays`. Appreciate your time!

Electronic-Waste reviewed (commented):
Basically LGTM!

Commit:
Update manifest: make namespace configurable

Signed-off-by: Xinmin Du <10803082+doris-xm@user.noreply.gitee.com>
Signed-off-by: Xinmin Du <2812493086@qq.com>

Electronic-Waste reviewed (commented):
Thanks for your great contribution and your patience! 

I think, we should also add namespace configuration here:

https://github.com/kubeflow/training-operator/blob/1dfa40c12516fc9eb2ce12c5ef52da7d46670457/manifests/v2/overlays/only-manager/kustomization.yaml#L1-L3

https://github.com/kubeflow/training-operator/blob/1dfa40c12516fc9eb2ce12c5ef52da7d46670457/manifests/v2/overlays/standalone/kustomization.yaml#L1-L3

to install these components in `kubeflow-system` namespace.

Commit:
Update manifest: move standalone, only-manager installation in namespace: kubeflow-system

Signed-off-by: Xinmin Du <10803082+doris-xm@user.noreply.gitee.com>
Signed-off-by: Xinmin Du <2812493086@qq.com>

Electronic-Waste reviewed (approved):
LGTM! Thanks for updating this:)

/lgtm
/assign @kubeflow/wg-training-leads @kubeflow/wg-manifests-leads @saileshd1402

andreyvelich commented on manifests/v2/base/manager/kustomization.yaml:
```diff
@@ -1,4 +1,2 @@
 resources:
   - manager.yaml
-# TODO (andreyvelich): Move it to overlays once we copy the JobSet manifests.
-namespace: kubeflow-system
```
How this is going to work since we don't copy JobSet manifests in the Kubeflow Training ?

Doris-xm replied:
Currently, the JobSet manifest is referred from a remote URL, so the `namespace` configuration in the base or overlays cannot override the JobSet manifests, I think. 

From the changlog, it seems that the latest release hasn’t included the changes in [kubernetes-sigs/jobset/#719](https://github.com/kubernetes-sigs/jobset/pull/719). We may need to manually modify the namespace setting for JobSet. I'm not sure if I need to deal with it in this pr.

andreyvelich replied:
Recently @kannon92 made a PR to support other namespaces in JobSet: https://github.com/kubernetes-sigs/jobset/pull/719.
Probably, we should use `kubeflow` namespace to install the JobSet operator within Kubeflow Platform if @kubeflow/wg-training-leads or @kubeflow/wg-manifests-leads don't have any other suggestions.

Electronic-Waste replied:
However, this PR seems not to be included in the current release of jobset:

https://github.com/kubernetes-sigs/jobset/compare/release-6.0...v0.7.2

In the manifest, the namespace is hardcoded into jobset-system：

https://github.com/kubernetes-sigs/jobset/releases/download/v0.7.2/manifests.yaml

Electronic-Waste replied:
@kannon92 I would appreciate it if you could provide some guide on how to install jobset components on custom namespace❤️

kannon92 replied:
Yes we don't have a release yet with that.

kannon92 replied:
@andreyvelich what is the timeline for this? I was thinking that the next release would contain your DependsOn work and the above PR. 

We could maybe make a case that the namespace one is an important bug fix (@ahg-g) and include it in a patch release.

andreyvelich replied:
@kannon92 It would be nice to have it before manifest testing ends for Kubeflow 1.10 on Feb 10th: https://github.com/kubeflow/community/tree/master/releases/release-1.10#:~:text=Monday%2C%20Feb%2010th%202025%20%3A%20Week%2017%20%2D%20Manifests%20sync%20ends%2C%20manifests%20testing%20and%20bug%20fixing%20starts.


@Electronic-Waste @kubeflow/wg-training-leads @Doris-xm Any suggestions on how we should include JobSet manifests into Kubeflow Training kustomize manifests to configure namespace name ?
Do we need to copy the entire manifests into Kubeflow Training GitHub repo ?

andreyvelich replied:
cc @kubeflow/wg-manifests-leads @kubeflow/release-team

Electronic-Waste replied:
Copying manifest file, from my perspective, is an option with low priority since we need to track the update of it every time it is released and copy all of them to training-operator again.

My suggestion would be:

1. Use envsubset(easiest way but need support from jobset team),like:
```
export NAMESPACE=my-dynamic-namespace
```
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-deployment
  namespace: ${NAMESPACE}
spec:
  replicas: 1
```
```

envsubst < deployment.yaml | kubectl apply -f -

```

2. PatchMergeStrategy

3. Vars and ReplacementTransformers

And it will also be nice to hear some more elegant way to install jobset components in custom namespaces from @kannon92 :)

WDYT👀 @Doris-xm @kubeflow/wg-training-leads @kubeflow/wg-manifests-leads

kannon92 replied:
I don’t think there is an elegant way AFAIK. If you use cert rotation then it needs to be Jobset-system. The PR linked would fix that. I can see about a cherry pick for that. If I forgot please feel free to propose it.

andreyvelich replied:
I think, the most straightforward approach is to copy JobSet manifests into Kubeflow Training and rely on the specific release of JobSet. We should make an appropriate changes to the JobSet config to deploy manager in the `kubeflow-system` namespace. We can always bump the JobSet version since it is very simple update.

In the future, we can design the automation script, similar to what @juliusvonkohout did for Kubeflow Manifests: https://github.com/kubeflow/manifests/blob/master/hack/synchronize-training-operator-manifests.sh

> I can see about a cherry pick for that. If I forgot please feel free to propose it.

@kannon92 Let me see if need to get this change before JobSet v0.8 (e.g. v0.7.3).
I would love to get the DependsOn API in JobSet v0.8 as well, since the MPI Runtime will use it.

Electronic-Waste replied:
@andreyvelich Do you mean that we should copy JobSet manifests to `standalone` and `kubeflow-platform` dir and change the namespace to `kubeflow-system` and `kubeflow` respectively? Is it a bit of redundant?

WDYT👀 @Doris-xm @kubeflow/wg-training-leads @kubeflow/wg-manifests-leads

andreyvelich replied:
We don't need to copy the entire manifests twice, the namespace can be controlled via [kustomize overlay](https://github.com/kubeflow/training-operator/blob/master/manifests/overlays/kubeflow/kustomization.yaml#L2).
E.g. for the standalone overlay, we will use `kubeflow-system`, for `kubeflow-platform`, we will use `kubeflow`.

Additionally, we have this open issue to remove namespace parameter from the JobSet config and read it from the Pod's filesystem: https://github.com/kubernetes-sigs/jobset/issues/720.

Electronic-Waste replied:
SGTM. For me, I just don't know how to overlay namespace items outside of the `.metadata` field, like:

<img width="663" alt="截屏2025-01-14 10 28 31" src="https://github.com/user-attachments/assets/5931bf1f-6544-4d2f-8b28-1aeaa6aac457" />

andreyvelich replied:
We can use the var references, similar to [this example](https://github.com/kubeflow/training-operator/blob/master/manifests/base/webhook/kustomizeconfig.yaml#L3-L7). We will tell Kustomize how to override the namespace value.

Electronic-Waste replied:
SGTM. That's what I want to do with the JobSet manifests!

Electronic-Waste replied:
@Doris-xm So now we can wait for the next release of JobSet. And make some appropriate changes to the manifest  after that:)

Doris-xm replied:
Thank you all for the insightful discussion! I’ve learned a lot from this. Much appreciated!

Doris-xm committed:
Merge branch 'refs/heads/master' into add-overlay-manifest-v2

# Conflicts:
#	manifests/v2/base/manager/kustomization.yaml
#	manifests/v2/base/rbac/kustomization.yaml
#	manifests/v2/base/webhook/kustomization.yaml
#	manifests/v2/overlays/only-manager/kustomization.yaml
#	manifests/v2/overlays/standalone/kustomization.yaml

Doris-xm committed:
Update manifest: add overlay for Kubeflow Platform installation

Signed-off-by: Xinmin Du <2812493086@qq.com>

Electronic-Waste reviewed (commented):
Basically LGTM. PTAL if you have time @kubeflow/wg-training-leads @kubeflow/wg-manifests-leads @astefanutti 

Btw, can you mark some addressed comments as `resolved` @Doris-xm ? Some of them are outdated.

/lgtm

juliusvonkohout commented:
> Basically LGTM. PTAL if you have time @kubeflow/wg-training-leads @kubeflow/wg-manifests-leads @astefanutti
> 
> Btw, can you mark some addressed comments as `resolved` @Doris-xm ? Some of them are outdated.
> 
> /lgtm

I think when you create the PR against kubeflow/manifests including the integrations tests to see how it behaves with the platform components and authorization, security etc. we can provide more input.

andreyvelich commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,23 @@
+  - ../../base/manager
+  - ../../base/rbac
+  - ../../base/webhook
+  - ../../base/runtimes/pretraining
```
The runtimes must be installed after controller-manager is ready.
Are we ok with that ?

astefanutti replied:
I fear this is going to be "annoying" for every downstream projects.

Maybe a "sub-optimal" solution would be to create the preset / in-tree runtimes before the webhooks, similar to what the [default "legacy" sorting](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/sortoptions/#example-3-default-legacy-sorting) does.
But the webhooks will have to be forward compatible during upgrades.

Otherwise we move towards CEL based validation or ValidatingAdmissionPolicy, which may be more appropriate/suitable for (Cluster)TrainingRuntime resources.

andreyvelich replied:
>Maybe a "sub-optimal" solution would be to create the preset / in-tree runtimes before the webhooks, similar to what the [default "legacy" sorting](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/sortoptions/#example-3-default-legacy-sorting) does.

Unfortunately, even with sorting Kustomize don't for resource probs before deploying the next resources.

astefanutti replied:
No I meant creating the runtimes before the webhooks.

andreyvelich replied:
Oh, I see.
That might work. @tenzen-y @Electronic-Waste Any thoughts ?

Doris-xm replied:
Currently, I try to solve it by modifying the order:
```yaml
resources:
  - ../../base/crds
  - ../../base/manager
  - ../../base/rbac
  - ../../base/runtimes/pretraining
  - ../../base/webhook
  - ../../third-party/jobset # Comment this line if JobSet is installed on the Kubernetes cluster.
  - kubeflow-trainer-roles.yaml
```
not sure if this will deploy as expected. Look forward to your further guidance.

astefanutti replied:
That might be safer to set the sorting strategy to FIFO:

```yaml
kind: Kustomization
sortOptions:
  order: fifo
```

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,84 @@
+  - apiGroups:
+      - ""
+    resources:
+      - persistentvolumeclaims
```
This is not required for now, please remove it.

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,84 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: kubeflow-trainer-edit
```
I think, we also should have permission to read logs from TrainJob's pods.

Doris-xm committed:
add permission for pods log read & rm persistentvolumeclaims

Signed-off-by: Xinmin Du <2812493086@qq.com>

Doris-xm committed:
create the runtimes before the webhooks

Signed-off-by: Xinmin Du <2812493086@qq.com>

Doris-xm committed:
Specify sorting order: fifo

Signed-off-by: Xinmin Du <2812493086@qq.com>

Doris-xm commented:
Thanks for your detailed reviews. Appreciate for your patience and guidance.
I've addressed the comments above. PTAL if you have time:) @kubeflow/wg-training-leads

Electronic-Waste commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,25 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: kubeflow
+sortOptions:
+  order: fifo
+resources:
+  - ../../base/crds
+  - ../../base/manager
+  - ../../base/rbac
+  - ../../base/runtimes/pretraining
```
Does FIFO work fine to install all these components? Could you examine it and provide the screenshot here? @Doris-xm

Doris-xm replied:
Thanks for suggestion. I will create a PR in `kubeflow/manifests` to test it under CI/CD.
For the deployment, it's installed in the following order:
<details>
<summary> kubectl apply --server-side -f kubeflow-platform1.yaml </summary>
customresourcedefinition.apiextensions.k8s.io/clustertrainingruntimes.trainer.kubeflow.org serverside-applied
customresourcedefinition.apiextensions.k8s.io/trainingruntimes.trainer.kubeflow.org serverside-applied
customresourcedefinition.apiextensions.k8s.io/trainjobs.trainer.kubeflow.org serverside-applied
deployment.apps/kubeflow-trainer-controller-manager serverside-applied
service/kubeflow-trainer-controller-manager serverside-applied
clusterrole.rbac.authorization.k8s.io/kubeflow-trainer-controller-manager serverside-applied
clusterrolebinding.rbac.authorization.k8s.io/kubeflow-trainer-controller-manager serverside-applied
serviceaccount/kubeflow-trainer-controller-manager serverside-applied
validatingwebhookconfiguration.admissionregistration.k8s.io/validator.trainer.kubeflow.org serverside-applied
customresourcedefinition.apiextensions.k8s.io/jobsets.jobset.x-k8s.io serverside-applied
serviceaccount/jobset-controller-manager serverside-applied
role.rbac.authorization.k8s.io/jobset-leader-election-role serverside-applied
clusterrole.rbac.authorization.k8s.io/jobset-manager-role serverside-applied
clusterrole.rbac.authorization.k8s.io/jobset-metrics-reader serverside-applied
clusterrole.rbac.authorization.k8s.io/jobset-proxy-role serverside-applied
rolebinding.rbac.authorization.k8s.io/jobset-leader-election-rolebinding serverside-applied
clusterrolebinding.rbac.authorization.k8s.io/jobset-manager-rolebinding serverside-applied
clusterrolebinding.rbac.authorization.k8s.io/jobset-proxy-rolebinding serverside-applied
secret/jobset-webhook-server-cert serverside-applied
service/jobset-controller-manager-metrics-service serverside-applied
service/jobset-webhook-service serverside-applied
deployment.apps/jobset-controller-manager serverside-applied
mutatingwebhookconfiguration.admissionregistration.k8s.io/jobset-mutating-webhook-configuration serverside-applied
validatingwebhookconfiguration.admissionregistration.k8s.io/jobset-validating-webhook-configuration serverside-applied
configmap/jobset-manager-config serverside-applied
clusterrole.rbac.authorization.k8s.io/kubeflow-trainer-edit serverside-applied
clusterrole.rbac.authorization.k8s.io/kubeflow-trainer-view serverside-applied
secret/kubeflow-trainer-webhook-cert serverside-applied
</details>

Howerver, I will examine it further these days.

Electronic-Waste replied:
Amazing! That sounds good to me. I have no other comment. Do you have any other ideas? @kubeflow/wg-training-leads @astefanutti

astefanutti replied:
One thing is that, as a required dependency, logically JobSet should be deployed first, so maybe move it in first position of the list, WDYT?

Electronic-Waste replied:
> One thing is that, as a required dependency, logically JobSet should be deployed first, so maybe move it in first position of the list, WDYT?

SGTM

Electronic-Waste reviewed (commented):
Sorry for the late review @Doris-xm. Basically LGTM. Thanks!

> I think when you create the PR against kubeflow/manifests including the integrations tests to see how it behaves with the platform components and authorization, security etc. we can provide more input.

It will be better if you could adopt the suggestions from @juliusvonkohout and create another PR againset `kubeflow/manifests`: https://github.com/kubeflow/manifests/issues/2948

/cc @kubeflow/wg-training-leads @astefanutti @kubeflow/wg-manifests-leads

Doris-xm committed:
Deploy jobset first

Signed-off-by: Xinmin Du <2812493086@qq.com>

andreyvelich commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -4,12 +4,12 @@ namespace: kubeflow
 sortOptions:
   order: fifo
```
Should we do the same for the standalone installation ?
So we don't need to install manager + runtimes in the separate overlays.

Doris-xm replied:
Thanks for suggestion. I start to work on it.

Electronic-Waste commented:
Thanks for this @Doris-xm! I believe, we can move forward with this PR. In the future, we can raise another PR to integrate trainer into the kubeflow platform: https://github.com/kubeflow/manifests/issues/2948

/lgtm
/assign @kubeflow/wg-training-leads @astefanutti

juliusvonkohout commented:
Hello, can we merge this @andreyvelich 
Afterwards we can continue in kubeflow/manifests and check how a simple training job as integration tests can be used similar to https://github.com/kubeflow/manifests/blob/1534a3bf02fecc39b41a41f9f7853666ef62de53/.github/workflows/training_operator_test.yaml#L69

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,86 @@
+      - trainingruntimes
+      - trainjobs
+    verbs:
+      - create
+      - delete
+      - get
+      - list
+      - patch
+      - update
+      - watch
```
Actually, I think `kubeflow-trainer-edit` doesn't require to Create, Update, or Delete runtimes, they only need to:
`get`, `list`, `watch`. using the Kubeflow SDK.

andreyvelich commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,25 @@
+  - ../../base/crds
+  - ../../base/manager
+  - ../../base/rbac
+  - ../../base/runtimes/pretraining
```
Shall we install runtimes right after the `../../base/crds` since we don't need to wait for `manager` to start ?

andreyvelich replied:
TODO: We should move our runtimes from the `pretraining` folder.

andreyvelich commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,25 @@
+  - kubeflow-trainer-roles.yaml
+
+# Update the Kubeflow Trainer controller manager image tag.
+images:
```
One think that comes to mind is that we also should control the version of runtimes we maintain:
DeepSpeed runtime, MLX Runtime.

cc @Electronic-Waste @astefanutti @kubeflow/wg-training-leads

andreyvelich replied:
Any thoughts on this @Doris-xm ?

Electronic-Waste replied:
@andreyvelich Do you mean that we should also have versions like `v1alpha1`, `v1` and more? Or define the version according to the release version like `v2.0.0`?

I think, maybe we do not need to maintain these versions since users can just install specific version of trainer with:

```
kubectl apply --server-side -k "github.com/kubeflow/trainer.git/mainfests/overlays/kubeflow-platform?ref=v2.0.0
```

Then, we'll use the `v2.0.0` version of runtimes. Does this sound good to you? @andreyvelich @Doris-xm

andreyvelich replied:
I meant that we should tag the images that we will release as part of Kubeflow Trainer release:
```
docker pull ghcr.io/kubeflow/trainer/trainer-controller-manager:v2.0.0
docker pull ghcr.io/kubeflow/trainer/mlx-runtime:v2.0.0
docker pull ghcr.io/kubeflow/trainer/torchtune-trainer:v2.0.0
docker pull ghcr.io/kubeflow/trainer/model-initializer:v2.0.0
docker pull ghcr.io/kubeflow/trainer/dataset-initializer:v2.0.0
docker pull ghcr.io/kubeflow/trainer/deepspeed-runtime:v2.0.0
```

Then, after user installs one of the overlays, we should use the correct version of the controller + runtimes.

@Electronic-Waste @Doris-xm Can we update our runtime overlay to include image version for the runtimes ?
https://github.com/kubeflow/trainer/blob/master/manifests/overlays/runtimes/kustomization.yaml

andreyvelich replied:
Yes, for the release branch, we should tag the images with the release tag.
@milinddethe15 Should work on this as part of this PR: https://github.com/kubeflow/trainer/pull/2623.

@milinddethe15 Do you think you have enough time to finalize this release script this or early next week ?

Doris-xm replied:
Does this requirement mean that we need to patch the image fields in the CRDs in the runtime overlay to override the image versions? @andreyvelich

andreyvelich replied:
Yeah, we probably should create patch to update the runtime images.
WDYT @astefanutti @kubeflow/wg-training-leads @Electronic-Waste ?

Electronic-Waste replied:
SGTM. But we need tag the images with the release tag first: https://github.com/kubeflow/trainer/pull/2623

andreyvelich replied:
For now, we can just apply patch with the `latest`tag, similar to the controller: https://github.com/kubeflow/trainer/blob/master/manifests/overlays/manager/kustomization.yaml#L18

Electronic-Waste replied:
SGTM

Doris-xm committed:
remove edit permissions to runtimes; install runtimes after crds

Signed-off-by: Xinmin Du <2812493086@qq.com>

Doris-xm committed:
Merge branch 'refs/heads/master' into add-overlay-manifest-v2

Doris-xm committed:
remove pretraining directory

Signed-off-by: Xinmin Du <2812493086@qq.com>

andreyvelich commented:
Hi @Electronic-Waste @Doris-xm, can someone finish this PR, so we can include it in the Kubeflow Trainer 2.0 release please ?

Electronic-Waste commented:
@andreyvelich I think, this PR is pending due to https://github.com/kubeflow/trainer/pull/2527 is unfinished. Maybe we can review it first:)

Doris-xm committed:
patch runtimes images

Signed-off-by: Xinmin Du <2812493086@qq.com>

Doris-xm commented:
Can we merge this PR now with the `latest` image tag ？ @andreyvelich @Electronic-Waste

Electronic-Waste commented:
> Can we merge this PR now with the latest image tag ？

I agree. I think we can move forward with this PR after we merge https://github.com/kubeflow/trainer/pull/2527

andreyvelich commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,29 @@
+
+# Update the Kubeflow Trainer controller manager & runtimes image tag.
+images:
+  - name: kubeflow/trainer-controller-manager
```
This should have correct image: https://github.com/kubeflow/trainer/blob/3658d1c0ee6bad1a7de3f15703b9d51807ed5920/manifests/overlays/manager/kustomization.yaml#L17

andreyvelich commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,29 @@
+# Update the Kubeflow Trainer controller manager & runtimes image tag.
+images:
+  - name: kubeflow/trainer-controller-manager
+    newTag: latest
+  - name: ghcr.io/kubeflow/trainer/deepspeed-runtime
+    newTag: latest
+  - name: ghcr.io/kubeflow/trainer/mlx-runtime
+    newTag: latest
```
@Doris-xm Did you verify that this Kustomize override is going to work ?

andreyvelich replied:
I noticed that @Electronic-Waste already tested it, so it looks good.

Electronic-Waste replied:
I think, it's just the same as https://github.com/kubeflow/trainer/pull/2590#discussion_r2081001181

andreyvelich commented on manifests/overlays/runtimes/kustomization.yaml:
```diff
@@ -2,3 +2,10 @@ apiVersion: kustomize.config.k8s.io/v1beta1
   - ../../base/runtimes
+
+# Update the Kubeflow Trainer runtimes image tag.
+images:
```
We also need to patch the torch tune Trainer image cc @Electronic-Waste

Doris-xm commented:
/rerun

Doris-xm commented:
/rerun all

Doris-xm commented:
/rerun-all

Doris-xm committed:
fix: correct image

Signed-off-by: Xinmin Du <2812493086@qq.com>

Doris-xm committed:
add image patch for more runtimes

Signed-off-by: Xinmin Du <2812493086@qq.com>

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,93 @@
+    rbac.authorization.kubeflow.org/aggregate-to-kubeflow-trainer-admin: "true"
+rules:
+  - apiGroups:
+      - kubeflow.org
```
I think, we should use correct apiGroup here
```suggestion
      - trainer.kubeflow.org
```

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,93 @@
+      - list
+      - watch
+  - apiGroups:
+      - kubeflow.org
```
```suggestion
      - trainer.kubeflow.org
```

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,93 @@
+      - update
+      - watch
+  - apiGroups:
+      - kubeflow.org
```
```suggestion
      - trainer.kubeflow.org
```

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,93 @@
+    rbac.authorization.kubeflow.org/aggregate-to-kubeflow-view: "true"
+rules:
+  - apiGroups:
+      - kubeflow.org
```
```suggestion
      - trainer.kubeflow.org
```

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,93 @@
+      - list
+      - watch
+  - apiGroups:
+      - kubeflow.org
```
```suggestion
      - trainer.kubeflow.org
```

Doris-xm committed:
Update manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml

Co-authored-by: Andrey Velichkevich <andrey.velichkevich@gmail.com>
Signed-off-by: Du Xinmin <2812493086@qq.com>

Doris-xm committed:
Update manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml

Co-authored-by: Andrey Velichkevich <andrey.velichkevich@gmail.com>
Signed-off-by: Du Xinmin <2812493086@qq.com>

Doris-xm committed:
Update manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml

Co-authored-by: Andrey Velichkevich <andrey.velichkevich@gmail.com>
Signed-off-by: Du Xinmin <2812493086@qq.com>

Doris-xm committed:
Merge branch 'master' into add-overlay-manifest-v2

Signed-off-by: Du Xinmin <2812493086@qq.com>

Doris-xm committed:
Update manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml

Co-authored-by: Andrey Velichkevich <andrey.velichkevich@gmail.com>
Signed-off-by: Du Xinmin <2812493086@qq.com>

Doris-xm committed:
Update manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml

Co-authored-by: Andrey Velichkevich <andrey.velichkevich@gmail.com>
Signed-off-by: Du Xinmin <2812493086@qq.com>

andreyvelich commented on manifests/overlays/kubeflow-platform/kubeflow-trainer-roles.yaml:
```diff
@@ -0,0 +1,93 @@
+  - apiGroups:
+      - trainer.kubeflow.org
+    resources:
+      - clustertrainingruntimes
```
I noticed that this is not going to work since we can't grant cluster-scoped resources via RoleBinding.
@juliusvonkohout @Doris-xm @kimwnasptd @thesuperzapper @Electronic-Waste Any thoughts on how we can grant access to ClusterTrainingRuntime in Kubeflow Notebooks ?

juliusvonkohout replied:
Only with a clusterrolebinding and that should be read only and very strict.

andreyvelich replied:
Yes, this is read only since we don't allow ML Users to modify the cluster training runtime.
I noticed that we have other cluster-scoped resources defined in `default-editor` rolebinding that just don't work.
Any thoughts on these and how we want to handle it with Profile Controller ?

juliusvonkohout replied:
The same is true then for kserve servingruntimes. So far it was good enough for users to not see them directly then.

Electronic-Waste replied:
> Any thoughts on how we can grant access to ClusterTrainingRuntime in Kubeflow Notebooks ?
Any thoughts on these and how we want to handle it with Profile Controller ?

Do you mean in `kubeflow/manifests`? Could you elaborate a bit?

I noticed that ClusterRoleBinding is used in this repo:

- Notebook App: https://github.com/kubeflow/manifests/blob/533ca9f8c1c97836fae6f5946c76ce31f42dd2c9/apps/jupyter/notebook-controller/upstream/rbac/role_binding.yaml#L1-L11
- Profile App: https://github.com/kubeflow/manifests/blob/533ca9f8c1c97836fae6f5946c76ce31f42dd2c9/apps/profiles/upstream/rbac/role_binding.yaml#L1-L11

> The same is true then for kserve servingruntimes. So far it was good enough for users to not see them directly then.

AFAIK, users may need to see the ClusterTrainingRuntimes(CTRs) in Notebook App or Profile Controller. We have increasing number of CTRs, which are hard to be remembered by users. Also, it will be more straightforward for users to get the CTRs using `TrainerClient.list_runtimes()` and reference it, like what we wrote in the get-started example in our official doc: 

```
from kubeflow.trainer import TrainerClient, CustomTrainer

for r in TrainerClient().list_runtimes():
    print(f"Runtime: {r.name}")

```

andreyvelich replied:
Yes, reading CTR is required to create Kubeflow TrainJobs. Without it, user doesn't know what runtimes are available for them to work with.

In the future, users should be able to get runtime packages, like what we've showed here: https://youtu.be/Fnb1a5Kaxgo?t=560

Doris-xm replied:
> users may need to see the ClusterTrainingRuntimes(CTRs) in Notebook App or Profile Controller.

Maybe we can use rolebinding to allow Notebook & Profile Controller to read CTRs. I wonder how I can verify this feature and test my code.

andreyvelich commented:
@Electronic-Waste @Doris-xm Given that we still working on this PR, can we cut the Kubeflow Trainer 2.0 release and include this change in the Kubeflow Trainer 2.1 given that Kubeflow Manifests release will be ~ September/October ?

Electronic-Waste commented:
Maybe we could try to fix it in recent days. Hey @Doris-xm, do you get a chance to finish this PR?

Doris-xm commented:
I’ll give it a try and do my best to get it done within a week. @Electronic-Waste 
cc @andreyvelich

Doris-xm committed:
Merge branch 'refs/heads/master' into add-overlay-manifest-v2

Doris-xm committed:
role_bind for notebook & profile

Signed-off-by: Xinmin Du <2812493086@qq.com>

Doris-xm committed:
Merge branch 'refs/heads/master' into add-overlay-manifest-v2

andreyvelich commented on manifests/overlays/runtimes/kustomization.yaml:
```diff
@@ -5,6 +5,10 @@ resources:
 
 # Update the Kubeflow LLM Trainer image tag.
 images:
+  - name: ghcr.io/kubeflow/trainer/deepspeed-runtime
+    newTag: latest
+  - name: ghcr.io/kubeflow/trainer/mlx-runtime
+    newTag: latest
```
Not in the master branch, we should keep them `v2.0.0-rc.0` on the `release-2.0` branch.

Doris-xm committed:
fix: reorder images

Signed-off-by: Xinmin Du <2812493086@qq.com>

Doris-xm commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,35 @@
+kind: Kustomization
+namespace: kubeflow
+sortOptions:
+  order: fifo
```
`fifo` works in a Kubernetes cluster.
I conducted the experiment and found that after removing the fifo flag, the following error occurred:

> Error from server (InternalError): Internal error occurred: failed calling webhook "validator.clustertrainingruntime.trainer.kubeflow.org": failed to call webhook: Post "https://kubeflow-trainer-controller-manager.kubeflow.svc:443/validate-trainer-kubeflow-org-v1alpha1-clustertrainingruntime?timeout=10s": no endpoints available for service "kubeflow-trainer-controller-manager"

Doris-xm committed:
fix: reuse overlay/manager & runtimes

Signed-off-by: Xinmin Du <2812493086@qq.com>

andreyvelich commented on manifests/overlays/kubeflow-platform/kustomization.yaml:
```diff
@@ -0,0 +1,7 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: kubeflow
```
You might need to create patch to remove this namespace:https://github.com/kubeflow/trainer/blob/7335204c3db0611672d5c82848a9a86c8dabca6b/manifests/overlays/manager/namespace.yaml

Otherwise, we can create Kubeflow Trainer resources in `kubeflow-system` namespace.

andreyvelich replied:
We need to make sure that namespace is created by Kubeflow Manifests: https://github.com/kubeflow/manifests/blob/master/example/kustomization.yaml#L62
Since it contains the appropriate labels for other tools (e.g. Katib).
Thus, we should remove the namespace creation as part of Kubeflow Trainer manifests for Kubeflow Platform.

You can see example in Katib: https://github.com/kubeflow/katib/blob/master/manifests/v1beta1/installs/katib-with-kubeflow/kustomization.yaml#L41

Doris-xm committed:
fix: remove namespace with patch

Signed-off-by: Xinmin Du <2812493086@qq.com>

andreyvelich reviewed (commented):
Thank you for this contribution @Doris-xm!
/lgtm
/approve
cc @juliusvonkohout @kubeflow/wg-manifests-leads @akagami-harsh you can integrate Kubeflow Trainer in Kubeflow Manifests

Electronic-Waste reviewed (approved):
@Doris-xm Thanks for this amazing work!
/lgtm
/approve

andreyvelich commented:
/cherry-pick release-2.0

juliusvonkohout commented:
@kunal-511 can you synchronize Trainer 2.0 to Kubeflow/manifests ?

kunal-511 commented:
> @kunal-511 can you synchronize Trainer 2.0 to Kubeflow/manifests ?

okay i will do that
